Enable dot notation in parameter expressions

smokyabdulrahman opened the pull request:
to resolve #44 

please ignore committed tests and dependencies, they were added to showcase:
- support of dot notation
- support of dicts and objects traversal

tests will be fixed once solution approach is validated.

maldoinc commented:
Hi, thanks for this. I think it generally looks fine. I'll give it a more thorough review in a few days as I also want to think a bit more what parameters look like.

We should not raise the minimum version just for match though.

smokyabdulrahman commented:
> Hi, thanks for this. I think it generally looks fine. I'll give it a more thorough review in a few days as I also want to think a bit more what parameters look like.
>
I will be awaiting your inputs.
 
> We should not raise the minimum version just for match though.
You are right, I just did that to test with Pydantic which requires 3.9 minimum. This will be reverted once the main logic is sorted out, and proper unit/integration tests are added.

smokyabdulrahman commented on wireup/ioc/parameter.py:
```diff
@@ -46,12 +46,32 @@ def __get_value_from_name(self, name: str) -> Any:
+
+                return holder[name]
+            case object():
+                if getattr(holder, name, None) is None:
```
I think here maybe I should be more careful.
what if the attribute exists but has `None` as value? I believe in this case I should return None (the value of the attribute)

smokyabdulrahman replied:
Then I think the code is doing just that. It is also throwing if `baz` doesn't exist. does that make sense?
example:
```json
{
  "foo": {
    "bar": {}
  }
}
```

maldoinc commented on test/integration/fastapi/router.py:
```diff
@@ -36,8 +36,13 @@ async def rng_route(random_service: Injected[RandomService]):
 @router.get("/params")
-async def params_route(foo: Annotated[str, Inject(param="foo")], foo_foo: Annotated[str, Inject(expr="${foo}-${foo}")]):
-    return {"foo": foo, "foo_foo": foo_foo}
+async def params_route(
```
We can add these directly as unit tests to the parameterbag class.

smokyabdulrahman replied:
I am just testing pydantic settings injection with this change. If you think it's not needed (doesn't add value) I can remove it.

smokyabdulrahman commented:
@maldoinc addressed your comments.
There is 1 test that broke after the changes: https://github.com/maldoinc/wireup/blob/6d62250ef0c5f3792e34a6857c6cbf952a016208/test/unit/test_container_core.py#L77-L93

What should be done do you think? (replaced the '.' with '_' in my PR for now for this test)

smokyabdulrahman commented:
any updates regarding this?

maldoinc commented:
Since there was a recent release of 2.0 I think it's a bit early to do another breaking change, but the change looks good.

To get around this I we can do the following: If a parameter exists with the given name (even if it contains dots) we can return that, otherwise perform nested resolution as necessary.

Apologies for taking so long, just needed to find a way around this.

maldoinc commented:
@smokyabdulrahman let me know if you want to drag this to the finish line.

smokyabdulrahman commented:
> @maldoinc I think all is done now.

Removed `match` statement to support `3.8`

maldoinc commented on wireup/ioc/parameter.py:
```diff
@@ -46,12 +46,39 @@ def __get_value_from_name(self, name: str) -> Any:
 
+    @classmethod
+    def __get_value_from_name_and_holder(cls, name: str, holder: Any) -> Any:
+        if isinstance(holder, dict):
```
suggestion: `isinstance(holder, Mapping)`

maldoinc commented on test/unit/test_parameter.py:
It'd be great to have some tests for the following:

* Testing that the priority is given to existing names even if they contain dots: foo.bar given `{"foo.bar": 1, "foo": {"bar": 2}}`
* Malformed paths: "a..b", ".foo", "foo."
* All the tests use parameter expressions, would be good to just use `.get` on the name directly without interpolation. `.get("db.host")`.
* object with a `@property`

maldoinc commented:
This is now released in 2.6.0
